lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -1417,95 +1517,108 @@ struct HandshakeCanonicalizePattern : public ConversionPattern {
-    rewriter.applySignatureConversion(&newFuncOp.getBody(), result);
+LogicalResult lowerFuncOp(mlir::FuncOp funcOp, MLIRContext *ctx) {
+  // Only retain those attributes that are not constructed by build.
+  SmallVector<NamedAttribute, 4> attributes;
+  for (const auto &attr : funcOp->getAttrs()) {
+    if (attr.first == SymbolTable::getSymbolAttrName() ||
+        attr.first == function_like_impl::getTypeAttrName())
+      continue;
+    attributes.push_back(attr);
+  }
 
-    // Remove redundant start op
-    Block *entryBlock = &newFuncOp.front();
-    // After signature conversion, control value is set as last entry block
-    // argument
-    Value val = entryBlock->getArguments().back();
-    Operation *cntrlMg = getStartOp(entryBlock);
-    cntrlMg->getResult(0).replaceAllUsesWith(val);
-    rewriter.eraseOp(cntrlMg);
+  // Get function arguments
+  llvm::SmallVector<mlir::Type, 8> argTypes;
+  for (auto &arg : funcOp.getArguments()) {
+    mlir::Type type = arg.getType();
+    argTypes.push_back(type);
+  }
 
-    rewriter.eraseOp(funcOp);
+  // Get function results
+  llvm::SmallVector<mlir::Type, 8> resTypes;
+  for (auto arg : funcOp.getType().getResults())
+    resTypes.push_back(arg);
+
+  handshake::FuncOp newFuncOp;
+
+  // Add control input/output to function arguments/results and create a
+  // handshake::FuncOp of appropriate type
+  wrapRewriter<mlir::FuncOp>(
+      [&](mlir::FuncOp funcOp, ConversionPatternRewriter &rewriter) {
+        auto noneType = rewriter.getNoneType();
+        argTypes.push_back(noneType);
+        resTypes.push_back(noneType);
+        auto func_type = rewriter.getFunctionType(argTypes, resTypes);
+        newFuncOp = rewriter.create<handshake::FuncOp>(
+            funcOp.getLoc(), funcOp.getName(), func_type, attributes);
+        rewriter.inlineRegionBefore(funcOp.getBody(), newFuncOp.getBody(),
+                                    newFuncOp.end());
+        return success();
+      },
+      ctx)(funcOp);
+
+  // Rewrite affine.for operations.
+  if (failed(wrapRewriter<handshake::FuncOp>(rewriteAffineFor, ctx)(newFuncOp)))
+    return newFuncOp.emitOpError("failed to rewrite Affine loops");
+
+  // Perform dataflow conversion
+  MemRefToMemoryAccessOp MemOps;
+  wrapRewriter<handshake::FuncOp>(
+      [&](handshake::FuncOp nfo, ConversionPatternRewriter &rewriter) {
+        MemOps = replaceMemoryOps(nfo, rewriter);
+        return success();
+      },
+      ctx)(newFuncOp);
+
+  wrapRewriter<handshake::FuncOp>(replaceCallOps, ctx)(newFuncOp);
+  wrapRewriter<handshake::FuncOp>(setControlOnlyPath, ctx)(newFuncOp);
+  wrapRewriter<handshake::FuncOp>(addMergeOps, ctx)(newFuncOp);
+  wrapRewriter<handshake::FuncOp>(addBranchOps, ctx)(newFuncOp);
+  wrapRewriter<handshake::FuncOp>(addSinkOps, ctx)(newFuncOp);
+  wrapRewriter<handshake::FuncOp>(connectConstantsToControl, ctx)(newFuncOp);
+  wrapRewriter<handshake::FuncOp>(addForkOps, ctx)(newFuncOp);
```

Comment:
Why the explicit template argument?  I think template deduction should work here?

Reply:
Also since you always seem to immediately call the result of wrapRewriter, why bother with the std::function?

Reply:
Unfortunately not, template argument deduction fails without explicitly stating the type.

> Also since you always seem to immediately call the result of wrapRewriter, why bother with the std::function?

Correct, that is redundant. I guess the current design is mainly due to the semantics of `wrapRewriter` (i.e., wrapping the function). I have no problem with removing the additional `std::function` step - `wrapRewriter` should then be renamed. Any suggestions?